compose-test.bash

```diff
@@ -0,0 +1,276 @@
+    export KEYSTORE_PASS
+fi
+
+child_process() {
+    while true; do
+        sleep 10
+        echo "Scaling down scaled-app..."
+        if [ "$CONTAINER_ENGINE" = "podman" ]; then
+            podman-compose kill scaled-app
+        else 
+            docker compose scale scaled-app=0
+        fi 
+
+        sleep 5
+
+        echo "Stopping stopped-app..."
+        $COMPOSE_ENGINE stop stopped-app
+
+        sleep 10
+
+        echo "Scaling up scaled-app..."
+        if [ "$CONTAINER_ENGINE" = "podman" ]; then
+            podman-compose restart scaled-app
+        else 
+            docker compose scale scaled-app=1
+        fi
+
+        sleep 5
+
+        echo "Restarting stopped-app..."
+        $COMPOSE_ENGINE restart stopped-app
+    done
+}
+
+# options post-handling
+
+# testing automated rule
+if [ "$CT_EN_RULES" = true ]; then
+    # generate automated-rule that matches vertx-fib-demos
+    touch "$(dirname "$0")"/cryostat-compose-test.json
+    echo "{\"name\":\"cryostat-compose-test\",\"description\":\"\",\"matchExpression\":\"target.alias == 'es.andrewazor.demo.Main'\",\"eventSpecifier\":\"template=Continuous,type=TARGET\",\"archivalPeriodSeconds\":30,\"initialDelaySeconds\":30,\"preservedArchives\":3,\"maxAgeSeconds\":35,\"maxSizeBytes\":0,\"enabled\":true}" >> "$(dirname "$0")"/cryostat-compose-test.json
+    # move to conf.d
+    mv "$(dirname "$0")"/cryostat-compose-test.json "$(dirname "$0")"/conf/rules/cryostat-compose-test.json
+else 
+    # delete that file if it exists
+    rm -f "$(dirname "$0")"/conf/rules/cryostat-compose-test.json
+fi
+
+# FIXME: podman-compose does not support COMPOSE_PROFILES yet https://github.com/containers/podman-compose/pull/592
+# testing invalid targets
+
+PROFILE_ARGS=""
+if [ "$CT_EN_INVALID" = true ]; then
+    # COMPOSE_PROFILES+=(",invalid")
+    PROFILE_ARGS+="--profile invalid "
+fi
+
+# testing duplicate Cryostat instance
+if [ "$CT_EN_DUPLICATE" = true ]; then
+    # COMPOSE_PROFILES+=(",duplicate")
+    PROFILE_ARGS+="--profile duplicate "
+fi
+
+# trap on CTRL+C SIGINT (we don't want restart signals to tear down the test environment)
+trap cleanup INT TERM
+
+# export COMPOSE_PROFILES
+
+merged_yaml="---
+services:"
+
+sections=("vertx-jmx" "quarkus-agent")
+replicas_variables=("CT_JMX_REPLICAS" "CT_AGENT_REPLICAS")
+
+non_zero_replicas=false
+
+# Iterate over the sections
+for ((i=0; i<${#sections[@]}; i++))
+do
+    section="${sections[i]}"
+    replicas_var="${replicas_variables[i]}"
+    replicas="${!replicas_var}"
+
+    if (( replicas > 0 )); then
+        non_zero_replicas=true
+
+        for ((j=1; j<=replicas; j++))
+        do
+            current_yaml="$(sed "s/$section/$section-${j}/g" "compose-$section.yaml")"
+            current_yaml="${current_yaml//---/}"; current_yaml="${current_yaml//services:/}"
+            merged_yaml+="$current_yaml"
+        done
+    fi
+done
+
+if $non_zero_replicas; then
+    echo "$merged_yaml" > compose-merged-tmp.yaml
```

Comment:
Maybe this file should be written to `/tmp`? ex. using `mktemp`.

Or if not, then perhaps just add a cleanup function that removes it on exit?

Reply:
Fixed by writing a cleanup function.